controllers/hcloudmachine_controller.go

```diff
@@ -355,3 +374,103 @@ func IgnoreHetznerClusterConditionUpdates(logger logr.Logger) predicate.Funcs {
 		GenericFunc: func(_ event.GenericEvent) bool { return true },
 	}
 }
+
+// IgnoreInsignificantMachineStatusUpdates is a predicate used for ignoring insignificant Machine.Status updates.
+func IgnoreInsignificantMachineStatusUpdates(logger logr.Logger) predicate.Funcs {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			log := logger.WithValues(
+				"predicate", "IgnoreInsignificantMachineStatusUpdates",
+				"type", "update",
+				"namespace", e.ObjectNew.GetNamespace(),
+				"kind", strings.ToLower(e.ObjectNew.GetObjectKind().GroupVersionKind().Kind),
+				"name", e.ObjectNew.GetName(),
+			)
+
+			var oldMachine, newMachine *clusterv1.Machine
+			var ok bool
+			// This predicate only looks at Machine objects
+			if oldMachine, ok = e.ObjectOld.(*clusterv1.Machine); !ok {
+				return true
+			}
+			if newMachine, ok = e.ObjectNew.(*clusterv1.Machine); !ok {
+				// Something weird happened, and we received two different kinds of objects
+				return true
+			}
+
+			// We should not modify the original objects, this causes issues with code that relies on the original object.
+			oldMachine = oldMachine.DeepCopy()
+			newMachine = newMachine.DeepCopy()
+
+			oldMachine.ManagedFields = nil
+			newMachine.ManagedFields = nil
+
+			oldMachine.ResourceVersion = ""
+			newMachine.ResourceVersion = ""
```

Comment:
Is this actually a problem or is this filtered out by controller-runtime beforehand? Feels like it _should_ be ignored by controller-runtime.

Reply:
I just copy pasted the code from you :D didn't think more about it

Reply:
Did a quick look through the controller-runtime code, but did not see any explicit handling of these fields, so we probably need to consider them here.

For managed fields there might be a solution in `controller-runtime` 0.18.0 (currently on 0.17.x). Someone added a cache transformer to remove these fields from the object cache. IIUC this cache is also used to evaluate the reconcile enqueing, so objects would be passed to the predicates without the managed fields, but this should be tested: https://github.com/kubernetes-sigs/controller-runtime/pull/2791


```patch
diff --git a/main.go b/main.go
index 4e8269ab..7a2a0321 100644
--- a/main.go
+++ b/main.go
@@ -123,6 +123,7 @@ func main() {
 			ByObject:          secretutil.AddSecretSelector(),
 			SyncPeriod:        &syncPeriod,
 			DefaultNamespaces: watchNamespaces,
+			DefaultTransform:  cache.TransformStripManagedFields(),
 		},
 	})
 	if err != nil {
```

Reply:
sounds interesting! We should keep this in mind for the CAPI 1.8 upgrade